### ISSUE TITLE
Support more than 28 interfaces per appliance by adding PCI bridges

### ISSUE THREAD

**Issue by arnegoetje** [GNS3/gns3-server#889]
Currently GNS3 only supports up to 28 network interfaces per appliance.
This is due to the current design using a single default PCI bridge and attaching all devices to that PCI bridge.

As explained here, it's possible to add multiple (up to 9) PCI bridges with up to 32 devices each:
https://github.com/qemu/qemu/blob/master/docs/pcie.txt 

The 32 device limitation is the PCI design. There are 32 devices per PCI bridge. The VM needs a few PCI devices for non-network related stuff, hence 28 interfaces can still be used on the default bridge. 

I've tried the following with a Cumulus VX VM:

By default GNS3 puts all the network interfaces under the single default PCI bridge: 

arne@GNS3:~$ ps axw | grep Edge1-SiteA 
45613 ? Sl 2:33 /usr/bin/qemu-system-x86_64 -name Edge1-SiteA -m 384M -smp cpus=1 -enable-kvm -machine smm=off -boot order=c -drive file=/home/arne/GNS3/projects/cebdc187-83c8-481d-b4a5-6470661abd08/project-files/qemu/9ac362ef-602b-437b-8b33-e0f061998654/hda_disk.qcow2,if=ide,index=0,media=disk -serial telnet:0.0.0.0:5000,server,nowait -monitor tcp:127.0.0.1:33554,server,nowait -net none -device virtio-net-pci,mac=00:bd:08:86:54:00 -device virtio-net-pci,mac=00:bd:08:86:54:01,netdev=gns3-1 -netdev socket,id=gns3-1,udp=10.100.45.228:10000,localaddr=10.100.45.228:10001 -device virtio-net-pci,mac=00:bd:08:86:54:02,netdev=gns3-2 -netdev socket,id=gns3-2,udp=10.100.45.228:10002,localaddr=10.100.45.228:10003 -device virtio-net-pci,mac=00:bd:08:86:54:03,netdev=gns3-3 -netdev socket,id=gns3-3,udp=10.100.45.228:10013,localaddr=10.100.45.228:10012 -device virtio-net-pci,mac=00:bd:08:86:54:04,netdev=gns3-4 -netdev socket,id=gns3-4,udp=10.100.45.228:10014,localaddr=10.100.45.228:10015 -device virtio-net-pci,mac=00:bd:08:86:54:05 -device virtio-net-pci,mac=00:bd:08:86:54:06 -device virtio-net-pci,mac=00:bd:08:86:54:07 -device virtio-net-pci,mac=00:bd:08:86:54:08 -device virtio-net-pci,mac=00:bd:08:86:54:09 -device virtio-net-pci,mac=00:bd:08:86:54:0a -device virtio-net-pci,mac=00:bd:08:86:54:0b -device virtio-net-pci,mac=00:bd:08:86:54:0c -device virtio-net-pci,mac=00:bd:08:86:54:0d -device virtio-net-pci,mac=00:bd:08:86:54:0e -device virtio-net-pci,mac=00:bd:08:86:54:0f -nographic 

On the VM this looks like: 

```
root@edge1-sitea:/home/cumulus# lspci 
00:00.0 Host bridge: Intel Corporation 440FX - 82441FX PMC [Natoma] (rev 02) 
00:01.0 ISA bridge: Intel Corporation 82371SB PIIX3 ISA [Natoma/Triton II] 
00:01.1 IDE interface: Intel Corporation 82371SB PIIX3 IDE [Natoma/Triton II] 
00:01.3 Bridge: Intel Corporation 82371AB/EB/MB PIIX4 ACPI (rev 03) 
00:02.0 VGA compatible controller: Device 1234:1111 (rev 02) 
00:03.0 Ethernet controller: Red Hat, Inc Virtio network device 
00:04.0 Ethernet controller: Red Hat, Inc Virtio network device 
00:05.0 Ethernet controller: Red Hat, Inc Virtio network device 
00:06.0 Ethernet controller: Red Hat, Inc Virtio network device 
00:07.0 Ethernet controller: Red Hat, Inc Virtio network device 
00:08.0 Ethernet controller: Red Hat, Inc Virtio network device 
00:09.0 Ethernet controller: Red Hat, Inc Virtio network device 
00:0a.0 Ethernet controller: Red Hat, Inc Virtio network device 
00:0b.0 Ethernet controller: Red Hat, Inc Virtio network device 
00:0c.0 Ethernet controller: Red Hat, Inc Virtio network device 
00:0d.0 Ethernet controller: Red Hat, Inc Virtio network device 
00:0e.0 Ethernet controller: Red Hat, Inc Virtio network device 
00:0f.0 Ethernet controller: Red Hat, Inc Virtio network device 
00:10.0 Ethernet controller: Red Hat, Inc Virtio network device 
00:11.0 Ethernet controller: Red Hat, Inc Virtio network device 
00:12.0 Ethernet controller: Red Hat, Inc Virtio network device 
```

Now when I add two more PCI bridges with two more interfaces connected to each: 

-device i82801b11-bridge,id=dmi_pci_bridge1 -device pci-bridge,id=pci-bridge1,bus=dmi_pci_bridge1,chassis_nr=0x1,addr=0x1,shpc=off -device virtio-net-pci,bus=pci-bridge1,addr=0x0,mac=00:bd:08:86:54:30 -device virtio-net-pci,bus=pci-bridge1,addr=0x1,mac=00:bd:08:86:54:31 -device pci-bridge,id=pci-bridge2,bus=dmi_pci_bridge1,chassis_nr=0x1,addr=0x2,shpc=off -device virtio-net-pci,bus=pci-bridge2,addr=0x0,mac=00:bd:08:86:54:40 -device virtio-net-pci,bus=pci-bridge2,addr=0x1,mac=00:bd:08:86:54:41 

This looks like the following on the VM: 

```
00:13.0 PCI bridge: Intel Corporation 82801 PCI Bridge (rev 92) 
01:01.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge 
01:02.0 PCI bridge: Red Hat, Inc. QEMU PCI-PCI bridge 
02:00.0 Ethernet controller: Red Hat, Inc Virtio network device 
02:01.0 Ethernet controller: Red Hat, Inc Virtio network device 
03:00.0 Ethernet controller: Red Hat, Inc Virtio network device 
03:01.0 Ethernet controller: Red Hat, Inc Virtio network device 
```

And the interfaces are available in the VM: 

```
swp16 Link encap:Ethernet HWaddr 00:bd:08:86:54:30 
inet6 addr: fe80::2bd:8ff:fe86:5430/64 Scope:Link 
UP BROADCAST RUNNING MULTICAST MTU:1500 Metric:1 
RX packets:0 errors:0 dropped:0 overruns:0 frame:0 
TX packets:279 errors:0 dropped:0 overruns:0 carrier:0 
collisions:0 txqueuelen:1000 
RX bytes:0 (0.0 B) TX bytes:56306 (54.9 KiB) 

swp17 Link encap:Ethernet HWaddr 00:bd:08:86:54:31 
inet6 addr: fe80::2bd:8ff:fe86:5431/64 Scope:Link 
UP BROADCAST RUNNING MULTICAST MTU:1500 Metric:1 
RX packets:0 errors:0 dropped:0 overruns:0 frame:0 
TX packets:279 errors:0 dropped:0 overruns:0 carrier:0 
collisions:0 txqueuelen:1000 
RX bytes:0 (0.0 B) TX bytes:56306 (54.9 KiB) 

swp18 Link encap:Ethernet HWaddr 00:bd:08:86:54:40 
inet6 addr: fe80::2bd:8ff:fe86:5440/64 Scope:Link 
UP BROADCAST RUNNING MULTICAST MTU:1500 Metric:1 
RX packets:0 errors:0 dropped:0 overruns:0 frame:0 
TX packets:279 errors:0 dropped:0 overruns:0 carrier:0 
collisions:0 txqueuelen:1000 
RX bytes:0 (0.0 B) TX bytes:56306 (54.9 KiB) 

swp19 Link encap:Ethernet HWaddr 00:bd:08:86:54:41 
inet6 addr: fe80::2bd:8ff:fe86:5441/64 Scope:Link 
UP BROADCAST RUNNING MULTICAST MTU:1500 Metric:1 
RX packets:0 errors:0 dropped:0 overruns:0 frame:0 
TX packets:279 errors:0 dropped:0 overruns:0 carrier:0 
collisions:0 txqueuelen:1000 
RX bytes:0 (0.0 B) TX bytes:56306 (54.9 KiB) 
```

So, what needs to be changed in GNS3 is the way it starts the qemu instances. 
It would need to add the PCI bridges first: 

```
-device i82801b11-bridge,id=dmi_pci_bridge1 \ 
-device pci-bridge,id=pci-bridge1,bus=dmi_pci_bridge1,chassis_nr=0x1,addr=0x1,shpc=off 
```

Then we can add up to 32 interfaces on that PCI bridge: 

```
-device virtio-net-pci,bus=pci-bridge1,addr=0x0,mac=00:bd:08:86:54:30 \ 
-device virtio-net-pci,bus=pci-bridge1,addr=0x1,mac=00:bd:08:86:54:31 \ 
... 
```

(The bus id and addr attribute are mandatory here) 

When the PCI bus is filled up, we can add another PCI bridge: 

```
-device pci-bridge,id=pci-bridge2,bus=dmi_pci_bridge1,chassis_nr=0x1,addr=0x2,shpc=off 
```

(The maximum number of PCI bridges is 9, as mentioned in the link above.) 

and add more interfaces: 

```
-device virtio-net-pci,bus=pci-bridge2,addr=0x0,mac=00:bd:08:86:54:40 \ 
-device virtio-net-pci,bus=pci-bridge2,addr=0x1,mac=00:bd:08:86:54:41 
... 
```

Of course the MAC addresses should follow the scheme set in GNS3. I've mocked them up for this test. 

In my test I've successfully added 64 data plane interfaces with two PCI bridges.

**Comment by noplay** [comment]
Well documented thanks for the issue !

**Comment by noplay** [comment]
It's merged for the 2.1 thanks a lot for the help
